controllers/operator/mongodbreplicaset_controller.go

```diff
@@ -76,58 +78,115 @@ type ReconcileMongoDbReplicaSet struct {
 	databaseNonStaticImageVersion     string
 }
 
+type ReplicaSetDeploymentState struct {
+	LastAchievedSpec *mdbv1.MongoDbSpec `json:"lastAchievedSpec"`
+}
+
 var _ reconcile.Reconciler = &ReconcileMongoDbReplicaSet{}
 
-func newReplicaSetReconciler(ctx context.Context, kubeClient client.Client, imageUrls images.ImageUrls, initDatabaseNonStaticImageVersion, databaseNonStaticImageVersion string, forceEnterprise bool, enableClusterMongoDBRoles bool, omFunc om.ConnectionFactory) *ReconcileMongoDbReplicaSet {
-	return &ReconcileMongoDbReplicaSet{
-		ReconcileCommonController: NewReconcileCommonController(ctx, kubeClient),
-		omConnectionFactory:       omFunc,
-		imageUrls:                 imageUrls,
-		forceEnterprise:           forceEnterprise,
-		enableClusterMongoDBRoles: enableClusterMongoDBRoles,
+// ReplicaSetReconcilerHelper contains state and logic for a SINGLE reconcile execution.
+// This object is NOT shared between reconcile invocations.
+type ReplicaSetReconcilerHelper struct {
+	resource        *mdbv1.MongoDB
+	deploymentState *ReplicaSetDeploymentState
+	reconciler      *ReconcileMongoDbReplicaSet
+	log             *zap.SugaredLogger
+}
 
-		initDatabaseNonStaticImageVersion: initDatabaseNonStaticImageVersion,
-		databaseNonStaticImageVersion:     databaseNonStaticImageVersion,
+func (r *ReconcileMongoDbReplicaSet) newReconcilerHelper(
+	ctx context.Context,
+	rs *mdbv1.MongoDB,
+	log *zap.SugaredLogger,
+) (*ReplicaSetReconcilerHelper, error) {
+	helper := &ReplicaSetReconcilerHelper{
+		resource:   rs,
+		reconciler: r,
+		log:        log,
+	}
+
+	if err := helper.initialize(ctx); err != nil {
+		return nil, err
 	}
+
+	return helper, nil
 }
 
-type deploymentOptionsRS struct {
-	agentCertPath        string
-	agentCertHash        string
-	prometheusCertHash   string
-	currentAgentAuthMode string
+// readState abstract reading the state of the resource that we store on the cluster between reconciliations.
+func (h *ReplicaSetReconcilerHelper) readState() (*ReplicaSetDeploymentState, error) {
+	// Try to get the last achieved spec from annotations and store it in state
+	if lastAchievedSpec, err := h.resource.GetLastSpec(); err != nil {
```

Comment:
this should probably one-off migration step?

Reply:
For now it's not a migration, I just abstracted the state reading/writing in a separate function (instead of randomly accessing annotations during the reconcile)
It will make more sense with a more complex state. I may even use the state_store at that point